action/protocol/staking/validations_test.go

```diff
@@ -155,3 +150,426 @@ func TestProtocol_ValidateCreateStake(t *testing.T) {
 		require.Equal(test.errorCause, errors.Cause(p.validateCreateStake(context.Background(), act)))
 	}
 }
+
+func TestProtocol_ValidateUnstake(t *testing.T) {
+	require := require.New(t)
+
+	p, _ := initTestProtocol(t)
+
+	tests := []struct {
+		bucketIndex uint64
+		payload     []byte
+		gasPrice    *big.Int
+		gasLimit    uint64
+		nonce       uint64
+		// expected results
+		errorCause error
+	}{
+		{
+			1,
+			[]byte("100000000000000000000"),
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			nil,
+		},
+		{1,
+			[]byte("100000000000000000000"),
+			big.NewInt(-unit.Qev),
+			10000,
+			1,
+			action.ErrGasPrice,
+		},
+	}
+
+	for _, test := range tests {
+		act, err := action.NewUnstake(test.nonce, test.bucketIndex, test.payload, test.gasLimit, test.gasPrice)
+		require.NoError(err)
+		require.Equal(test.errorCause, errors.Cause(p.validateUnstake(context.Background(), act)))
+	}
+	// test nil action
+	require.Equal(ErrNilAction, errors.Cause(p.validateUnstake(context.Background(), nil)))
+}
+
+func TestProtocol_ValidateWithdrawStake(t *testing.T) {
+	require := require.New(t)
+
+	p, _ := initTestProtocol(t)
+
+	tests := []struct {
+		bucketIndex uint64
+		payload     []byte
+		gasPrice    *big.Int
+		gasLimit    uint64
+		nonce       uint64
+		// expected results
+		errorCause error
+	}{
+		{
+			1,
+			[]byte("100000000000000000000"),
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			nil,
+		},
+		{1,
+			[]byte("100000000000000000000"),
+			big.NewInt(-unit.Qev),
+			10000,
+			1,
+			action.ErrGasPrice,
+		},
+	}
+
+	for _, test := range tests {
+		act, err := action.NewWithdrawStake(test.nonce, test.bucketIndex, test.payload, test.gasLimit, test.gasPrice)
+		require.NoError(err)
+		require.Equal(test.errorCause, errors.Cause(p.validateWithdrawStake(context.Background(), act)))
+	}
+	// test nil action
+	require.Equal(ErrNilAction, errors.Cause(p.validateWithdrawStake(context.Background(), nil)))
+}
+
+func TestProtocol_ValidateChangeCandidate(t *testing.T) {
+	require := require.New(t)
+
+	p, cands := initTestProtocol(t)
+
+	tests := []struct {
+		candName    string
+		bucketIndex uint64
+		payload     []byte
+		gasPrice    *big.Int
+		gasLimit    uint64
+		nonce       uint64
+		// expected results
+		errorCause error
+	}{
+		{
+			cands[0].Name,
+			1,
+			[]byte("100000000000000000000"),
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			nil,
+		},
+		{"12132323",
+			1,
+			[]byte("100000000000000000000"),
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
+		{"~1",
+			1,
+			[]byte("100000000000000000000"),
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
+		{"100000000000000000000",
+			1,
+			[]byte("100000000000000000000"),
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			ErrInvalidCanName,
+		},
```

Comment:
these 3 are actually duplicate case?

there are 4 error cases for validateChangeCandidate(), add test to cover them all